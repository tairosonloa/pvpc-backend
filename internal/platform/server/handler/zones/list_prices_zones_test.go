package zones

import (
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/gin-gonic/gin"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/require"

	pvpc "pvpc-backend/internal"
	"pvpc-backend/internal/listing"
	"pvpc-backend/internal/platform/storage/storagemocks"
)

func Test_ListZonesHandler_Success(t *testing.T) {
	gin.SetMode(gin.TestMode)
	repositoryMock := new(storagemocks.PricesZonesRepository)
	listingService := listing.NewListingService(repositoryMock)

	r := gin.New()
	r.GET("/zones", ListZonesHandler(listingService))

	zone1, err := pvpc.NewPricesZone(pvpc.PricesZoneDto{
		ID:         "ABC",
		ExternalID: "1234",
		Name:       "zone1",
	})
	require.NoError(t, err)

	zone2, err := pvpc.NewPricesZone(pvpc.PricesZoneDto{
		ID:         "DEF",
		ExternalID: "5678",
		Name:       "zone2",
	})
	require.NoError(t, err)

	repositoryMock.On(
		"GetAll",
		mock.Anything,
	).Return([]pvpc.PricesZone{zone1, zone2}, nil)

	req, err := http.NewRequest(http.MethodGet, "/zones", nil)
	require.NoError(t, err)

	rec := httptest.NewRecorder()
	r.ServeHTTP(rec, req)

	res := rec.Result()
	defer res.Body.Close()

	repositoryMock.AssertExpectations(t)
	require.Equal(t, http.StatusOK, res.StatusCode)
	snaps.MatchSnapshot(t, rec.Body.String())

}

func Test_ListZonesHandler_Empty(t *testing.T) {
	gin.SetMode(gin.TestMode)
	repositoryMock := new(storagemocks.PricesZonesRepository)
	listingService := listing.NewListingService(repositoryMock)

	r := gin.New()
	r.GET("/zones", ListZonesHandler(listingService))

	repositoryMock.On(
		"GetAll",
		mock.Anything,
	).Return([]pvpc.PricesZone{}, nil)

	req, err := http.NewRequest(http.MethodGet, "/zones", nil)
	require.NoError(t, err)

	rec := httptest.NewRecorder()
	r.ServeHTTP(rec, req)

	res := rec.Result()
	defer res.Body.Close()

	require.Equal(t, http.StatusOK, res.StatusCode)
	snaps.MatchSnapshot(t, rec.Body.String())
}

func Test_ListZonesHandler_Error(t *testing.T) {
	gin.SetMode(gin.TestMode)
	repositoryMock := new(storagemocks.PricesZonesRepository)
	listingService := listing.NewListingService(repositoryMock)

	r := gin.New()
	r.GET("/zones", ListZonesHandler(listingService))

	repositoryMock.On(
		"GetAll",
		mock.Anything,
	).Return(nil, errors.New("mock error"))

	req, err := http.NewRequest(http.MethodGet, "/zones", nil)
	require.NoError(t, err)

	rec := httptest.NewRecorder()
	r.ServeHTTP(rec, req)

	res := rec.Result()
	defer res.Body.Close()

	require.Equal(t, http.StatusInternalServerError, res.StatusCode)
	snaps.MatchSnapshot(t, rec.Body.String())

}
