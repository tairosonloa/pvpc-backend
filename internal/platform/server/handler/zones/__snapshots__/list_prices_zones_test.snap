
[Test_ListZonesHandler_Success - 1]
{"zones":[{"id":"ABC","external_id":"1234","name":"zone1"},{"id":"DEF","external_id":"5678","name":"zone2"}],"total":2}
---

[Test_ListZonesHandler_Empty - 1]
{"zones":[],"total":0}
---

[Test_ListZonesHandler_Error - 1]
{"error_code":"INTERNAL_SERVER_ERROR","message":"mock error","status_code":500}
---
